modules/nixos/disko-zfs-zvol.nix

```diff
@@ -0,0 +1,97 @@
+{
+  config,
+  inputs,
+  lib,
+  ...
+}:
+let
+  devices = idx: {
+    type = "disk";
+    device = "/dev/nvme${idx}n1";
+    content = {
+      type = "gpt";
+      partitions = {
+        boot = lib.mkIf (idx == "0") {
+          size = "1M";
+          type = "EF02"; # for grub MBR
+        };
+        ESP = lib.mkIf (idx == "0") {
+          size = "1G";
+          type = "EF00";
+          content = {
+            type = "filesystem";
+            format = "vfat";
+            mountpoint = "/boot";
+            mountOptions = [ "nofail" ];
+          };
+        };
+        zfs = {
+          size = "100%";
+          content = {
+            type = "zfs";
+            pool = "zroot";
+          };
+        };
+      };
+    };
+  };
+in
+{
+  imports = [ inputs.disko.nixosModules.disko ];
+
+  options = {
+    nixCommunity.disko.zramWritebackSize = lib.mkOption {
+      type = lib.types.int;
+      default = null;
+      description = "swap size in GB";
+    };
+  };
+
+  config = {
+    # this is both efi and bios compatible
+    boot.loader.grub = {
+      enable = true;
+      efiSupport = true;
+      efiInstallAsRemovable = true;
+    };
+
+    # Sometimes fails after the first try, with duplicate pool name errors
+    boot.initrd.systemd.services.zfs-import-zroot.serviceConfig.Restart = "on-failure";
+
+    zramSwap.writebackDevice = "/dev/zvol/zroot/writeback";
```

Comment:
Doesn't this suffer from the same issue that we also have with swap files on zfs filesystems? Also now that zfs has proper direct I/O maybe this is not an issue? https://github.com/openzfs/zfs/releases/tag/zfs-2.3.0-rc1

Reply:
zvol is used as an example in the zram-generator docs so I'm assuming that it works.

https://github.com/systemd/zram-generator/blob/766bfb0d297fe083cc29ab1886d8feff24f2ee58/man/zram-generator.conf.md?plain=1#L85